Features/cgu/24.2/various logical grid and tile grid improvements

cguglielmo opened the pull request.

cguglielmo committed:
LayoutConfig: work with real values

The LayoutConfig object did not contain the actual value unless
the values were set explicitly.
Instead, the values were read later by the layout which made it
impossible to work with these values before the widget was rendered.
With this change, the real values are not available on the config
object and can be accessed even if the widget is not rendered yet.

cguglielmo committed:
Tile: classId must not throw an error if there is no parent

This is necessary if a tile is created without a parent, e.g. if it is
created by the hybrid manager and used on a JS based tile grid.

cguglielmo committed:
TileGrid: add wrappable property

Some grids don't need the wrapping feature.
Disabling it improves unnecessary layouting operations which makes it a
little bit faster.

cguglielmo committed:
LogicalGrid: add property change event when gridData changes

It may be useful to detect whether the actual grid data changed.

To not trigger unnecessary events the GridData now implements the
equals method.

cguglielmo committed:
TileGrid: add isTileSelected method

cguglielmo committed:
TileGrid: allow disabling animateRemoval for specific tiles

Currently, remove animation could only be disabled for every tile in
the grid. But sometimes it is necessary to disable it for specific tiles
(e.g. custom placeholders).

cguglielmo committed:
TileGrid: animateTileInsertion cannot be disabled just for one operation

Sometimes, the animation should only be disabled for some operations.
E.g.:

this.tileGrid.animateTileInsertion = false;
this.tileGrid.setTiles(tiles);
this.tileGrid.animateTileInsertion = true;

This did not work because the flag was only checked later when the
layout animation was done and the flag already reset.

cguglielmo committed:
Tile: add animateBoundsChange property

Can be used to disable animation for size or position changes.

cguglielmo committed:
GridData: add clone method

This makes it easier to adjust grid data objects without modifying
the original object.

cguglielmo committed:
JsonTileField: fix generic

bschwarzent commented on eclipse-scout-core/src/layout/logicalgrid/GridData.ts:
```diff
@@ -168,6 +168,13 @@ export class GridData {
+  /**
+   * @returns a clone of this grid data enriched with the properties of the given model.
+   */
+  clone?(model?: GridData) {
```
- Declare the return type (GridData)

bschwarzent commented on eclipse-scout-core/src/tile/TileGridLayout.ts:
```diff
@@ -123,6 +123,9 @@ export class TileGridLayout extends LogicalGridLayout {
 
   protected _storeBounds(tiles: Tile[]) {
     tiles.forEach((tile, i) => {
+      if (!tile.animateBoundsChange) {
```
Seems okay, although it's not entirely clear how not setting some data properties disables the animation. I believe it relies on 6732927d0eb3f563e4a9e6080f1412770d37b4a1, correct?

cguglielmo replied:
you are right, the flag should be checked in _animateTiles as well

bschwarzent commented on eclipse-scout-core/src/tile/TileGrid.ts:
```diff
@@ -383,6 +383,9 @@ export class TileGrid<TTile extends Tile = Tile> extends Widget implements TileG
   }
 
   protected _renderInsertTiles(tiles: (TTile | PlaceholderTile)[]) {
+    if (!this.animateTileInsertion) {
```
Remark about the commit message of ee1a72a6779a66cff6159bdeb1d1bd55b126a1a3:
- Last line has an unnecessary space at the start.
- Do I understand the text correctly, that the example code _did not_ work, but now _does_ with this change?

bschwarzent commented on eclipse-scout-core/src/form/fields/FormField.ts:
```diff
@@ -847,10 +847,7 @@ export class FormField extends Widget implements FormFieldModel {
-      gridData = new GridData();
-    }
-    this._setProperty('gridDataHints', GridData.ensure(gridData));
+    this._setProperty('gridDataHints', GridData.ensure(gridData || new GridData()));
```
Why is this necessary: `|| new GridData()`
The method name "ensure" suggests that it automatically converts null or undefined arguments to a valid object (like arrays#ensure). Is that not the case?

cguglielmo replied:
It ensures the passed object has the GridData type, unless it is null or undefined. We use the same pattern for dates.ensure, numbers.ensure, LookupCall.ensure, DateFormat.ensure. I think this is correct because otherwise we would lose the null state which may be important. For arrays we explicitly don't want the null state. But I agree it is a little bit confusing...

bschwarzent commented on eclipse-scout-core/src/layout/logicalgrid/GridData.ts:
```diff
@@ -168,6 +168,13 @@ export class GridData {
     $.extend(this, model);
   }
 
+  equals?(other: GridData) {
```
- Declare the return type (boolean)
- What is the significance of the ? after the method name? If an implementation is provided with the base class, why should the method be declared as optional?

cguglielmo replied:
Until now, the GridData object contained no methods and could be used as model object. To prevent compile errors I added the ? modifier.

I agree, it is confusing. We should probably create a GridDataModel instead

bschwarzent commented on eclipse-scout-core/src/layout/logicalgrid/GridData.ts:
```diff
@@ -168,6 +168,13 @@ export class GridData {
   }
 
+  equals?(other: GridData) {
+    if (!other || !(other instanceof GridData)) {
```
First case is not necessary (undefined and null are also not instance of GridData)

bschwarzent commented on eclipse-scout-core/src/tile/TileGridLayout.ts:
```diff
@@ -217,7 +217,7 @@ export class TileGridLayout extends LogicalGridLayout {
     let fromBounds = tile.$container.data('oldBounds');
-    if (tile instanceof PlaceholderTile && !tile.$container.data('was-layouted')) {
+    if (!tile.$container.data('was-layouted')) {
       // Placeholders may not have fromBounds because they are added while layouting
```
Maybe adjust this comment  (it now seems to apply to all tiles, not only placeholder tiles)

bschwarzent commented on eclipse-scout-core/src/layout/logicalgrid/LogicalGridLayoutConfig.ts:
```diff
@@ -38,8 +38,8 @@ export interface LogicalGridLayoutConfigModel {
  *
- * The configured hints only have an effect if theirs value is >=0.
- * Otherwise, the default values specified by CSS are applied (see {@link LogicalGridLayout._initDefaults}).
+ * The configured hints only have an effect if their value is >=0.
```
Maybe use same condition in documentation than in code? (`>=0` vs. `> -1`)

bschwarzent commented on eclipse-scout-core/src/layout/logicalgrid/LogicalGridLayoutConfig.ts:
```diff
@@ -49,29 +49,47 @@ export class LogicalGridLayoutConfig implements LogicalGridLayoutConfigModel {
+  protected _options: InitModelOf<LogicalGridLayoutConfig>;
 
-  constructor(options?: InitModelOf<LogicalGridLayoutConfig>) {
+  constructor(options?: InitModelOf<LogicalGridLayoutConfig>, defaults?: InitModelOf<LogicalGridLayoutConfig>) {
```
There are many ways to contruct this object:

1. constructor - options + defaults
2. clone() - defaults only
3. ensure() - options only

Isn't this a bit confusing? When should which we use which method? There should be at least a little bit of documentation. I suspect that only the ensure() method should be called from outside, the other two are only internal, is that correct?

bschwarzent commented on eclipse-scout-core/src/layout/logicalgrid/LogicalGridLayoutConfig.ts:
```diff
@@ -102,4 +120,28 @@ export class LogicalGridLayoutConfig implements LogicalGridLayoutConfigModel {
+  }
+
+  static prepareSmallHgapConfig(layoutConfig?: ObjectOrModel<LogicalGridLayoutConfig>) {
+    if (!layoutConfig) {
```
Doesn't ensure() handle this case?

bschwarzent commented on eclipse-scout-core/src/layout/logicalgrid/LogicalGridLayoutConfig.ts:
```diff
@@ -102,4 +120,28 @@ export class LogicalGridLayoutConfig implements LogicalGridLayoutConfigModel {
+    });
+  }
+
+  static prepareSmallHgapConfig(layoutConfig?: ObjectOrModel<LogicalGridLayoutConfig>) {
```
Declare return type

bschwarzent replied:
Two widgets call ensure() to create the config object, and two other widgets call prepareSmallHgapConfig(). I think it would be better understandable if those two ways were more similar. Some ideas:
1. Add a defaults argument to ensure(). Usage:
`this._setProperty('layoutConfig', LogicalGridLayoutConfig.ensure(layoutConfig, {hgap: HtmlEnvironment.get().smallColumnGap}));`
2. Add a second ensureX() method. Usage:
`this._setProperty('layoutConfig', LogicalGridLayoutConfig.ensureWithSmallHgap(layoutConfig));`
4. Add a fluent method. Usage:
`this._setProperty('layoutConfig', LogicalGridLayoutConfig.ensure(layoutConfig).withSmallHgapGap());`

bschwarzent commented on eclipse-scout-core/src/layout/logicalgrid/LogicalGridLayoutConfig.ts:
```diff
@@ -102,4 +120,28 @@ export class LogicalGridLayoutConfig implements LogicalGridLayoutConfigModel {
     return new LogicalGridLayoutConfig(layoutConfig);
   }
+
+  static initHtmlEnvChangeHandler(widget: Widget, getter: () => LogicalGridLayoutConfig, setter: (config: LogicalGridLayoutConfig) => void) {
```
Maybe add a little documentation? The fact that this method must be called exactly once during the initialization of the widget is not clear without reading the code.

bschwarzent commented on eclipse-scout-core/src/tile/TileGridLayoutConfig.ts:
```diff
@@ -22,20 +22,42 @@ export interface TileGridLayoutConfigModel extends LogicalGridLayoutConfigModel
+
+  constructor(options?: InitModelOf<TileGridLayoutConfig>, defaults?: InitModelOf<TileGridLayoutConfig>) {
+    super(options, defaults);
+    this._init(); // Needs to be initialized again otherwise maxWidth would be set to undefined
```
Is that comment correct? $.extend() usually ignores properties that are undefined.

bschwarzent commented on eclipse-scout-core/src/form/fields/groupbox/GroupBox.ts:
```diff
@@ -283,6 +283,7 @@ export class GroupBox extends CompositeField implements GroupBoxModel {
       bodyLayoutConfig = new LogicalGridLayoutConfig();
     }
     this._setProperty('bodyLayoutConfig', LogicalGridLayoutConfig.ensure(bodyLayoutConfig));
+    LogicalGridLayoutConfig.initHtmlEnvChangeHandler(this, () => this.bodyLayoutConfig, layoutConfig => this.setBodyLayoutConfig(layoutConfig));
```
Remark about the commit message of e7f880c32249eae5f56823b88ee13d84f0d07e88:

> Default values are now available directly on the layout config object.
> This makes it easier to access the actual values for custom calculations.

What exactly does this mean? Is there an example for such a "custom calculation"? And why are the default values needed to access the "actual values"? Are these the widget-specific values or the final, merged values?

bschwarzent commented on eclipse-scout-core/src/tile/Tile.ts:
```diff
@@ -36,6 +36,8 @@ export class Tile extends Widget implements TileModel {
     this.selected = false;
     this.selectable = false;
     this.plainText = null;
+    // Null to let TileGrid decide whether to enable animation
```
Remark about the commit message of 6732927d0eb3f563e4a9e6080f1412770d37b4a1:
* There is an unnecessary space in front of the last line.

cguglielmo committed:
GridData: add model object to exclude equals and clone methods

The new methods must not be part of the model, so they don't appear
in code completion.
